File: packages/i18nify-go/modules/currency/is_valid_amount.go
package currency

import (
	"fmt"
	"regexp"
	"strconv"
	"strings"
)

var amountPattern = regexp.MustCompile(`^-?\d+(\.\d+)?$`)

// IsValidAmount reports whether amount has decimal precision within the
// allowed minor units for currencyCode. It returns an error when the
// currency code is not found in the dataset.
func IsValidAmount(amount string, currencyCode string) (bool, error) {
	currencyInfo, exists := cachedCurrencyData.CurrencyInformation[currencyCode]
	if !exists {
		return false, fmt.Errorf("invalid currency code: %s", currencyCode)
	}

	trimmed := strings.TrimSpace(amount)
	if !amountPattern.MatchString(trimmed) {
		return false, nil
	}

	allowedDecimals, err := strconv.Atoi(currencyInfo.MinorUnit)
	if err != nil {
		return false, fmt.Errorf("malformed minor_unit for currency %s", currencyCode)
	}

	dotIdx := strings.Index(trimmed, ".")
	actualDecimals := 0
	if dotIdx != -1 {
		actualDecimals = len(trimmed) - dotIdx - 1
	}

	return actualDecimals <= allowedDecimals, nil
}

File: packages/i18nify-js/src/modules/currency/isValidAmount.ts
import { withErrorBoundary } from '../../common/errorBoundary';
import CURRENCY_INFO from './data/currencyConfig.json';

const AMOUNT_PATTERN = /^-?\d+(\.\d+)?$/;

const isValidAmount = (amount: string, currencyCode: string): boolean => {
  if (typeof amount !== 'string' || typeof currencyCode !== 'string')
    return false;

  const trimmed = amount.trim();
  if (!AMOUNT_PATTERN.test(trimmed)) return false;

  if (!(currencyCode in CURRENCY_INFO)) return false;

  const { minor_unit } =
    CURRENCY_INFO[currencyCode as keyof typeof CURRENCY_INFO];
  const allowedDecimals = parseInt(minor_unit, 10);

  const dotIndex = trimmed.indexOf('.');
  const actualDecimals = dotIndex === -1 ? 0 : trimmed.length - dotIndex - 1;

  return actualDecimals <= allowedDecimals;
};

export default withErrorBoundary<typeof isValidAmount>(isValidAmount);
